### PR TITLE
fixing memory leak in copy constructor for forest class, and assignme…

### PR DESCRIPTION
**Pull request by shajoezhu**
…nt operator

### CODE REVIEW

**Comment by gerton** [comment]
Fingers crossed that this is the last one!

File: src/forest.h [hunk 1]
```diff
@@ -79,6 +79,7 @@ class Forest
 #endif
   friend class TimeInterval;
   friend class TimeIntervalIterator;
+  friend class TMRCA;
```
Comment (paulstaab) [inline src/forest.h]: Is this needed for the actual algorithm, or can it be moved up to the test setup?

**Comment by paulstaab** [comment]
Thanks!

File: src/summary_statistics/tmrca.cc [hunk 2]
```diff
@@ -22,13 +22,14 @@
 #include "tmrca.h"
 
 void TMRCA::calculate(const Forest &forest) {
-  if (forest.calcSegmentLength() == 0) return;
+  //std::cout << "forest.calcSegmentLength() "<<forest.calcSegmentLength() <<std::endl;
+  if (forest.rec_bases_.size() == 1) return;
```
Comment (shajoezhu) [inline src/summary_statistics/tmrca.cc]: Hi Paul, 

https://github.com/scrm/scrm/pull/92#discussion_r41728933, valgrind complained about for not initialized value here. Hence I made it friend to class forest.

Best
Joe

Comment (paulstaab) [inline src/summary_statistics/tmrca.cc]: Oh, I missed that change. The two `if`s do pretty different things here, don't they? Why was the original one wrong? I don't think that is fixes the uninitialized value. I'll look into it.

**Comment by paulstaab** [comment]
@shajoezhu Are you using multiple random generators in your particle filter or just one?

**Comment by shajoezhu** [comment]
Hi Paul

Our particle is an inherited class of forest, which we allow the forest to point to its own model and random generators (in order for us to do multi threading simulation, since we don't stdout, scrm work can work pretty well), by default we don't though, we just use one model, and one random generator.

best
joe

File: src/summary_statistics/tmrca.cc (hunk 2 above)
Comment (shajoezhu) [inline src/summary_statistics/tmrca.cc]: The first line wasn't wrong. When I was checking for the copy constructor and assignment operator, I was using valgrind on the unit_test, which complained about the uninitialised values on that line.

**Comment by paulstaab** [comment]
Hey again,

I think I found the problem, was again related to caching of contemporaries. Sorry for that! Copyconstructing Forest works pretty reliably now, I'm optimistic that it is now leak free.

I fixed the remaining memory leaks in the unittests (thanks for starting this @shajoezhu !). For me, the unittests now complete without any complains from valgrind.

Finally, I use a shared pointer for storing `FastFunc` inside of `RandomGenerator`. This might require some small type adjustments in the particle filter. Just replacing `FastFunc*` by `shared_ptr<FastFunc>` or better `auto` should most likely do the job. Also, the `clearFastFunc` is no longer needed can should be removed.

Best,
Paul

**Comment by shajoezhu** [comment]
Hi Paul

Thank you very much for this! All changes look pretty good.

Best
Joe

**Comment by paulstaab** [comment]
Command | Equal results | time master | time fixing_forest_copier | memory master | memory fixing_forest_copier
------- | ------------- | ----------- | -------- | ------------- | ---------
scrm 100 100 -t 5 -T | Yes | 0.092s | 0.091s | 3190kb | 3204kb
scrm 20 100 -r 10 100 -t 3 -T -L | Yes | 0.086s | 0.085s | 3168kb | 3179kb
scrm 4 1000 -r 10 100 -t 5 --transpose-segsites | Yes | 0.06s | 0.06s | 3265kb | 3273kb
scrm 20 2 -r 500 100 -L | Yes | 0.15s | 0.16s | 3205kb | 3211kb
scrm 2000 1 -r 10 100 -l 50 -L -T | Yes | 0.55s | 0.55s | 5998kb | 6004kb
scrm 20 1 -r 4000 10000000 -l 300000 -T | Yes | 0.89s | 0.9s | 4406kb | 4401kb
scrm 500 1 -r 1000 10000000 -l 0 -T | Yes | 0.63s | 0.63s | 3904kb | 3914kb
scrm 100 1 -r 400 10000000 -l 0 -O | Yes | 0.29s | 0.29s | 3209kb | 3207kb
scrm 10 100 -r 10 100 -I 3 3 3 4 0.5 -eN 0.1 0.05 -eN 0.2 0.5 -L | Yes | 0.054s | 0.054s | 3137kb | 3125kb
scrm 50 100 -r 20 200 -G 1.5 -l 100 -L | Yes | 0.047s | 0.047s | 3156kb | 3160kb

**Comment by shajoezhu** [comment]
Thanks Paul!

**Comment by gerton** [comment]
Great work everyone!!  Many thanks!

BW
Gerton

**Pull request merged**